kimchi/src/snarky/traits.rs

```diff
@@ -0,0 +1,136 @@
+//! Traits.
+
+use ark_ff::PrimeField;
+
+use super::{
+    checked_runner::{RunState, WitnessGeneration},
+    cvar::CVar,
+    prelude::TypeCreation,
+};
+
+/// A snarky type is a type that can be used in a circuit.
+/// It references an equivalent "out-of-circuit" type that one can use outside of the circuit.
+/// (For example, to construct private or public inputs, or a public output, to the circuit.)
+pub trait SnarkyType<F>: Sized
+where
+    F: PrimeField,
+{
+    /// Some additional information. (TODO: add more documentation.)
+    type Auxiliary;
+
+    /// The equivalent "out-of-circuit" type.
+    /// For example, the [super::boolean::Boolean] snarky type has an out-of-circuit type of [bool].
+    type OutOfCircuit;
+
+    /// The number of field elements that this type takes.
+    const SIZE_IN_FIELD_ELEMENTS: usize;
+
+    /// Returns the circuit variables (and auxiliary data) behind this type.
+    fn to_cvars(&self) -> (Vec<CVar<F>>, Self::Auxiliary);
+
+    /// Creates a new instance of this type from the given circuit variables (And some auxiliary data).
+    fn from_cvars_unsafe(cvars: Vec<CVar<F>>, aux: Self::Auxiliary) -> Self;
+
+    /// Checks that the circuit variables behind this type are valid.
+    /// For some definition of valid.
+    /// For example, a Boolean snarky type would check that the field element representing it is either 0 or 1.
+    /// The function does this by adding constraints to your constraint system.
+    fn check(&self, cs: &mut RunState<F>);
+
+    /// Deserializes this type into the out-of-circuit type (along with some auxiliary data).
+    fn deserialize(&self) -> (Self::OutOfCircuit, Self::Auxiliary);
+
+    /// Serializes the out-of-circuit type (along with some auxiliary data) into this type.
+    fn serialize(out_of_circuit: Self::OutOfCircuit, aux: Self::Auxiliary) -> Self;
+
+    /// Returns some auxiliary data (TODO: more doc).
+    fn constraint_system_auxiliary() -> Self::Auxiliary;
+
+    ///
+    fn value_to_field_elements(x: &Self::OutOfCircuit) -> (Vec<F>, Self::Auxiliary);
+
+    fn value_of_field_elements(x: (Vec<F>, Self::Auxiliary)) -> Self::OutOfCircuit;
+
+    //
+    // new functions that might help us with generics?
+    //
+
+    fn compute<FUNC>(
+        cs: &mut RunState<F>,
+        check: TypeCreation,
```

Comment:
It would be nice to call this `compute_generic` and have `compute`/`compute_unsafe` that specialize it (or equivalently remove this and have just those 2 functions). This API seems a little cumbersome for users.